File: 0739-daily-temperatures/0739-daily-temperatures.java
class Solution {
    public int[] dailyTemperatures(int[] temperatures) {
        int[] ans = new int[temperatures.length];
        Stack<Integer> st = new Stack<>();
        st.push(temperatures.length-1);
        for(int i=temperatures.length-2; i>=0; i--){
            if(temperatures[st.peek()]>temperatures[i]){
                ans[i] = st.peek()-i;
                st.push(i);
            }
            else{
                while(!(st.isEmpty()) && temperatures[st.peek()]<=temperatures[i]){
                    st.pop();
                }
                if(st.isEmpty()){
                    st.push(i);
                }
                else{
                    ans[i] = st.peek()-i;
                    st.push(i);
                }
            }
        }

        return ans;
    }
}

File: 0739-daily-temperatures/0739-daily-temperatures.py
class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        ans = [0]*len(temperatures)
        st = []
        st.append(len(temperatures)-1)
        for i in range(len(temperatures)-2, -1, -1):
            if(temperatures[st[-1]]>temperatures[i]):
                ans[i] = st[-1] - i
                st.append(i)
            else:
                while(not(len(st)==0) and temperatures[st[-1]]<=temperatures[i]):
                    st.pop()
                if(len(st)==0):
                    st.append(i)
                else:
                    ans[i] = st[-1] - i
                    st.append(i)
        
        return ans